File: builder/cpp/builder.cc
#include "builder/cpp/builder.h"
#include <string>

void Director::direct() {
    builder->buildpart1();
    builder->buildpart2();
    builder->buildpart3();
}

void Builder1::buildpart1() {
    result += "1";
}

void Builder1::buildpart2() {
    result += "2";
}

void Builder1::buildpart3() {
    result += "3";
}

std::string Builder1::get_result() {
    return result;
}

void Builder2::buildpart1() {
    result += 1;
}

void Builder2::buildpart2() {
    result += 2;
}

void Builder2::buildpart3() {
    result += 3;
}

int Builder2::get_result() {
    return result;
}

File: builder/builder.go
package builder

type Builder interface {
	Buildpart1()
	Buildpart2()
	Buildpart3()
}

type Director struct {
	b Builder
}

func NewDirector(b Builder) *Director {
	return &Director{
		b: b,
	}
}

func (d *Director) Construct() {
	d.b.Buildpart1()
	d.b.Buildpart2()
	d.b.Buildpart3()
}

type Builder1 struct {
	result string
}

func (b *Builder1) Buildpart1() {
	b.result += "1"
}

func (b *Builder1) Buildpart2() {
	b.result += "2"
}

func (b *Builder1) Buildpart3() {
	b.result += "3"
}

func (b *Builder1) GetResult() string {
	return b.result
}

type Builder2 struct {
	result int
}

func (b *Builder2) Buildpart1() {
	b.result += 1
}

func (b *Builder2) Buildpart2() {
	b.result += 2
}

func (b *Builder2) Buildpart3() {
	b.result += 3
}

func (b *Builder2) GetResult() int {
	return b.result
}
